src/master-playlist-controller.js

```diff
@@ -0,0 +1,519 @@
+      this.blacklistCurrentPlaylist_(this.masterPlaylistLoader_.error);
+    });
+
+    this.masterPlaylistLoader_.on('mediachanging', () => {
+      this.mainSegmentLoader_.pause();
+    });
+
+    this.masterPlaylistLoader_.on('mediachange', () => {
+      this.mainSegmentLoader_.abort();
+      this.mainSegmentLoader_.load();
+      this.hlsHandler.tech_.trigger({
+        type: 'mediachange',
+        bubbles: true
+      });
+    });
+
+    this.mainSegmentLoader_.on('progress', () => {
+      // figure out what stream the next segment should be downloaded from
+      // with the updated bandwidth information
+      this.hlsHandler.bandwidth = this.mainSegmentLoader_.bandwidth;
+      this.masterPlaylistLoader_.media(this.hlsHandler.selectPlaylist());
+
+      this.hlsHandler.tech_.trigger('progress');
+    });
+
+    this.mainSegmentLoader_.on('error', () => {
+      this.blacklistCurrentPlaylist_(this.mainSegmentLoader_.error());
+    });
+  }
+
+  load() {
+    this.mainSegmentLoader_.load();
+    if (this.audioPlaylistLoader_) {
+      this.audioSegmentLoader_.load();
+    }
+  }
+
+  loadAlternateAudioPlaylist(playlistLoader) {
+    this.audioPlaylistLoader_ = playlistLoader;
+
+    this.audioPlaylistLoader_.on('loadedmetadata', () => {
+      this.audioSegmentLoader_.playlist(this.audioPlaylistLoader_.media());
+
+      // if the video is already playing, or if this isn't a live video and preload
+      // permits, start downloading segments
+      if (!this.hlsHandler.tech_.paused() ||
+            (this.audioPlaylistLoader_.media().endList &&
+            this.hlsHandler.tech_.preload() !== 'metadata' &&
+            this.hlsHandler.tech_.preload() !== 'none')) {
+        this.audioSegmentLoader_.load();
+      }
+    });
+
+    this.audioPlaylistLoader_.on('loadedplaylist', () => {
+      let updatedPlaylist = this.audioPlaylistLoader_.media();
+      let seekable;
+
+      if (!updatedPlaylist) {
+        // only one playlist to select
+        this.audioPlaylistLoader_.media(
+          this.audioPlaylistLoader_.playlists.master.playlists[0]);
+        return;
+      }
+
+      this.audioSegmentLoader_.playlist(updatedPlaylist);
+    });
+
+    this.audioPlaylistLoader_.on('error', () => {
+      this.audioSegmentLoader_.abort();
+      this.audioPlaylistLoader_ = null;
+      // TODO go back to using combined
+    });
+
+    this.audioSegmentLoader_.on('error', () => {
+      this.audioSegmentLoader_.abort();
+      this.audioPlaylistLoader_ = null;
+      // TODO go back to using combined
+    });
+  }
+
+  /**
+   * Seek to the latest media position if this is a live video and the
+   * player and video are loaded and initialized.
+   */
+  setupFirstPlay() {
+    let seekable;
+    let media = this.masterPlaylistLoader_.media();
+
+    // check that everything is ready to begin buffering
+
+    // 1) the video is a live stream of unknown duration
+    if (this.hlsHandler.duration() === Infinity &&
```

Comment:
Many of the calls to `hlsHandler` in this method could be converted to queries against playlists or playlist loaders. `!this.mainPlaylistLoader_.media().endList` is equivalent to this comparison, for instance.